### PR TITLE
Issue/331 dialogs edits

### PR DESCRIPTION
**Pull request by vedmant**
#331

### CODE REVIEW

File: back/appsync/serverless.yml [hunk 1]
```diff
@@ -121,7 +121,7 @@ custom:
     number: 10
   pythonRequirements:
     dockerizePip: non-linux
-    pythonBin: /home/orion/envs/idea36/bin/python
+    pythonBin: python3
```
Comment (aldarund) [inline back/appsync/serverless.yml:124]: changing one hardcode to other :) not good. rather make it env variable than

Comment (vedmant) [reply back/appsync/serverless.yml:124]: For your env you can just make a python3 symlink in on of PATH folders. Probably we should create appropriate readme file with details.

File: front/package.json [hunk 2]
```diff
@@ -41,7 +41,8 @@
     "vue-advanced-cropper": "^0.16.0",
     "vue-notifications": "1.0.2",
     "vue-snotify": "3.2.1",
-    "vue-trix": "^0.9.0"
+    "vue-trix": "^0.9.0",
+    "vuex": "^3.4.0"
```
Comment (aldarund) [inline front/package.json:45]: Thats not correct. It will end up with multiple version of vuex used in the app e.g. current 3.1.1 that will be used for nuxt-vue-app and 3.4 global

File: back/appsync/serverless.yml (hunk 1 above)
Comment (aldarund) [reply back/appsync/serverless.yml:124]: why we need to do that if u can make this into env variable? And symlink is not something that will work properly for example on windows os

File: front/components/dialogs/DefaultDialog.vue [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    },
+    showCancelButton: {
+      type: Boolean,
+      default: true
+    }
+  }
+}
+</script>
+
+<style lang="scss">
```
Comment (aldarund) [inline front/components/dialogs/DefaultDialog.vue:81]: why not scoped

Comment (vedmant) [reply front/components/dialogs/DefaultDialog.vue:81]: Firstly is there any guidlines about scoping styles? Some places are scoped, others a not. Secondly, original component I refactored was not scoped https://github.com/ezl/dailyidea.com/blob/fdeb1b16b0790638ba3213dce76dbcbd8d28fc4c/front/components/dialogs/shareIdeaByEmail.vue#L192

File: front/package.json (hunk 2 above)
Comment (vedmant) [reply front/package.json:45]: Ok, removed it

File: back/appsync/serverless.yml (hunk 1 above)
Comment (vedmant) [reply back/appsync/serverless.yml:124]: Sure env variable is better, should we make a new ticket for this?

**Comment by ezl** [comment]
do it. for now just leave that part out and then create a ticket to turn
that into an env variable.

--Eric

File: front/components/dialogs/DefaultDialog.vue (hunk 3 above)
Comment (vedmant) [reply front/components/dialogs/DefaultDialog.vue:81]: There are list of places without scope https://prnt.sc/sh8kf0

Comment (aldarund) [reply front/components/dialogs/DefaultDialog.vue:81]: i know, it should be scoped everywhere its remains of some bad code which we have quite a lot

File: front/package.json (hunk 2 above)
Comment (aldarund) [reply front/package.json]: better ( proper way) way would be to update nuxt version and so it will also update vuex and other, but it might break some other things so better to do it in separate

File: front/components/dialogs/DefaultDialog.vue (hunk 3 above)
Comment (vedmant) [reply front/components/dialogs/DefaultDialog.vue:81]: I'll add it here, for other places should we create new issue

File: front/package.json (hunk 2 above)
Comment (vedmant) [reply front/package.json]: Yes, I just had some issue with IDE not finding vuex, so I added it as a fast fix, removed it already.

File: front/components/dialogs/DefaultDialog.vue (hunk 3 above)
Comment (vedmant) [reply front/components/dialogs/DefaultDialog.vue]: Ok, added "scoped"

**Comment by aldarund** [comment]
So instead of refs refactor passed var to the dialog and in every dialog we need to have it like this
>    @hide="() => (showFirstIdeaSaved = false)"
>    @hide="() => (showSavedByLoginLink = false)"
>    @hide="() => (showAskName = false)"
etc. So we have state in parent just for this even if we don't need it in most cases. Instead of self contained dialogs with refs :) And global bus. 
Basically its not worse solution, it is better, especially considering a lot of copy pate that was eliminated (  which could be done other way but nvm) , but its more complex. Harder to understand and use. And its for the sake of "calling directly from other component is absolutely not appropriate way to do things in Vue" ? Refs is appropriate way to do things in vue for something like dialog. Or form. For example look at one of the most used form validation package vee validate have it in docs too that if u need programmatic access - use refs. https://logaretm.github.io/vee-validate/guide/forms.html#programmatic-access-with-refs
TLDR - refs is appropriate usage in vue for specific cases. Otherwise it would be there (c) xD But since we already here and its refactored and this isnt really worse, but rather just more complex - let it be. 
But it also creates new issues.
For example as i see it don't clear field content on close and reopen, which is generally isn't the behaviour that is expected. This one need to be fixed and than its fine.

**Comment by vedmant** [comment]
@aldarund Yes, this is how it's supposed to be done, and done by everyone I've seen before. Vue is reactive framework, means it reacts on state change. What you were doing with refs is procedural programming, not reactive, that's completely opposite of reactive approach. If you look at your mentioned Vue bootstrap, you'll see exactly the same what I did (a bit more complex is it attaches itself to DOM). If you don't like `@hide="() => (showAskName = false)"` I can add v-model instead, that's not essential, v-model will basically do that same.

Programmatic access with refs is a workaround and exception for very rare occasions like this validator for example and even in Vee that's not main way of usage. In general practice calling Vue function of child component from parent component is absolute nonsense, only data passed down component and only events passed up components, that the most basic rule of Vue and any other reactive framework. 

In Vue docs using $refs is for handling edge cases https://vuejs.org/v2/guide/components-edge-cases.html 

> In most cases, it’s best to avoid reaching into other component instances or manually manipulating DOM elements. There are cases, however, when it can be appropriate.

modals are not that edge cases

File: front/components/dialogs/DefaultDialog.vue (hunk 3 above)
Comment (vedmant) [reply front/components/dialogs/DefaultDialog.vue]: I remove scoped cause I don't understand the point of adding shared styles:

```
<style scoped lang="scss">
@import '~assets/style/common';
@import '~assets/style/modals';
```

to every single component that uses dialog. It doesn't work if I only add styles to base dialog component. As I understand it will basically produce lots of duplicates of styles that otherwise are reused. So what's the point of 'scoped' here?

**Comment by aldarund** [comment]
>What's the point of clearing fields there?
the point is about general thing. It should be at least option 

>Programmatic access with refs is a workaround and exception for very rare occasions like this validator for example. In general practice calling Vue function of child component from parent component is not recomended, only data passed down component and only events passed up components, that the basic rule of Vue and any other reactive framework.

Lets one of the largest vue framework. Quasar
https://quasar.dev/quasar-plugins/dialog
They have exact this pattern
```
  show () {
      this.$refs.dialog.show()
    },

    // following method is REQUIRED
    // (don't change its name --> "hide")
    hide () {
      this.$refs.dialog.hide()
    },
```
So the thing is exactly same approach. And they find it totally fine. For one of the largest vue lib. 

>IWhat you were doing with refs is procedural programming, not reactive, that's completely opposite of reactive approach. 

What u mean not reactive? Refs are not reactive but they shouldn't be for this usecase.

File: front/components/dialogs/ResendAuthEmail.vue [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        >. Make sure the email address is spelled correctly, and try checking
+        your spam or junk folder. Click <strong>Back to Log In</strong> and
+        input correct email if you have mistaken.
+      </span>
+    </div>
+  </default-dialog>
+</template>
+
+<script>
+import * as Sentry from '@sentry/browser'
```
Comment (aldarund) [inline front/components/dialogs/ResendAuthEmail.vue:28]: we should do this. its importing from browser which could fail on ssr. And also nuxt sentry provide this.$sentry which should be used instead

**Comment by vedmant** [comment]
> Lets one of the largest vue framework. Quasar
https://quasar.dev/quasar-plugins/dialog
They have exact this pattern

That is one example, I can send you lots of examples where dialogs and modals are using with v-model. Plus your example is a dialog plugin (an alternative way), on dialog page they have usage examples with v-models everywhere and no any $refs example.

```js
<q-dialog v-model="alert">
      <q-card>
        <q-card-section>
```


Furthermore Vue docs specifically states that using $refs for accessing child component is an escape hatch:

> $refs are only populated after the component has been rendered, and they are not reactive. It is only meant as an escape hatch for direct child manipulation - you should avoid accessing $refs from within templates or computed properties.

This is from handling edge cases page https://vuejs.org/v2/guide/components-edge-cases.html

File: front/components/dialogs/ResendAuthEmail.vue (hunk 4 above)
Comment (vedmant) [reply front/components/dialogs/ResendAuthEmail.vue]: Ok, replaced with this.$sentry

**Comment by aldarund** [comment]
>That's not complexity, it's appropriate way of reactive programming, data goes down components tree, events go up, in case of $refs this is a violation as event (function call) start to do down.

https://forum.vuejs.org/t/the-right-way-to-open-a-modal-component/7013/2 see answer from vue core member than :)

**Comment by aldarund** [comment]
And what about
>the point is about general thing. It should be at least option to clear or not clear.

**Comment by vedmant** [comment]
> And what about

Not sure what option do you mean, if name and email have to be cleared they just can be cleared in parent component SaveIdeaBookmark, as this is parent component state, not dialog's. The question is do we need to do that?

**Comment by vedmant** [comment]
> https://forum.vuejs.org/t/the-right-way-to-open-a-modal-component/7013/2 see answer from vue core member than :)

That's 3 years old comment, also this is his IMO, we can ask other core dev what they think is the right way. In any case basically all of frameworks use v-model (or sync prop) as a main way in documentation:

- https://quasar.dev/vue-components/dialog
- https://vuetifyjs.com/en/components/dialogs/
- https://www.antdv.com/components/modal/
- https://vuematerial.io/components/dialog
- https://element.eleme.io/#/en-US/component/dialog
- https://www.iviewui.com/components/modal-en
- https://vuikit.js.org/guide/modal

 Do you thing they all are wrong?

**Comment by aldarund** [comment]
>Do you thing they all are wrong?

I never said its wrong to do it this way, I'm saying that ref isn't wrong too. Thats the point. Its fine to use refs, its used in real projects.

>That's 3 years old comment, also this is his IMO

ofc its his IMO. The point is that using ref is not WRONG AND SHOULDNT BE DONE THATS NOT HOW VUE SHOULD WORK xD

> (or sync prop)

Which is essentially data up which is > data goes down components tree, events go up, in case of $refs this is a violation as event (function call) start to do down.

violation :) Despite its wrapper it's still in the end data up
